pkg/controller/deployment/deployment_controller.go

```diff
@@ -22,79 +22,308 @@ import (
+				return dc.expClient.Deployments(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&extensions.Deployment{},
+		FullDeploymentResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: dc.enqueueDeployment,
+			UpdateFunc: func(old, cur interface{}) {
+				// Resync on deployment object relist.
+				dc.enqueueDeployment(cur)
+			},
+			// This will enter the sync loop and no-op, because the deployment has been deleted from the store.
+			// Note that deleting a controller immediately after scaling it to 0 will not work. The recommended
+			// way of achieving this is by performing a `stop` operation on the deployment.
+			DeleteFunc: dc.enqueueDeployment,
+		},
+	)
+
+	dc.rcStore.Store, dc.rcController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dc.client.ReplicationControllers(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return dc.client.ReplicationControllers(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.ReplicationController{},
+		ControllerRelistPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dc.addRC,
+			UpdateFunc: dc.updateRC,
+			DeleteFunc: dc.deleteRC,
+		},
+	)
+
+	// We do not event on anything from the podController, but we use the local
+	// podStore to make queries about the current state of pods (e.g. whether
+	// they are ready or not) more efficient.
+	dc.podStore.Store, dc.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dc.client.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return dc.client.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.Pod{},
+		PodRelistPeriod,
+		framework.ResourceEventHandlerFuncs{},
+	)
+
+	dc.syncHandler = dc.syncDeployment
+	return dc
 }
 
-func (d *DeploymentController) Run(syncPeriod time.Duration) {
-	go util.Until(func() {
-		errs := d.reconcileDeployments()
-		for _, err := range errs {
-			glog.Errorf("Failed to reconcile: %v", err)
-		}
-	}, syncPeriod, util.NeverStop)
+// When an RC is created, enqueue the deployment that manages it.
+func (dc *DeploymentController) addRC(obj interface{}) {
+	rc := obj.(*api.ReplicationController)
+	if d := dc.getDeploymentForRC(rc); rc != nil {
+		dc.enqueueDeployment(d)
+	}
 }
 
-func (d *DeploymentController) reconcileDeployments() []error {
-	list, err := d.expClient.Deployments(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+// getDeploymentForRC returns the deployment managing the given RC.
+// TODO: Surface that we are ignoring multiple deployments for a given controller.
```

Comment:
Can a replication controller be owned by more than one deployment? This means that the deployments are using the same pod template but have different names?